README.md
[![bswp crate](https://img.shields.io/crates/v/bswp.svg)](https://crates.io/crates/bswp)
[![bswp documentation](https://docs.rs/bswp/badge.svg)](https://docs.rs/bswp)
[![GitHub license](https://img.shields.io/github/license/PicoJr/bswp)](https://github.com/PicoJr/bswp/blob/master/LICENSE)

|Branch|Status|
|------|------|
|[master](https://github.com/PicoJr/bswp/tree/master)|![Build Status](https://github.com/PicoJr/bswp/workflows/Rust/badge.svg?branch=master)|

# BSWP

Rust Byte Swap lib. 

Swap bytes using patterns and masks.

## Minimum Supported Rust Version (MSRV)

`bswp` requires Rust [1.40.0](https://blog.rust-lang.org/2019/12/19/Rust-1.40.0.html).

## Usage

### Iterators

```rust
use bswp::pattern::{Pattern, Predicate, swap_iter};

let pattern = Pattern::new(0x42).with_mask(0xFF); // replace byte by 0x42
let predicate = Predicate::new().with_periodicity(2).with_offset(1); // replace odd bytes
let swaps = &[(pattern, predicate)];

let source: [u8; 4] = [0x41, 0x41, 0x41, 0x41];
let swapped = swap_iter(&source, swaps); // iterator on result
let swapped: Vec<u8> = swapped.collect();
assert_eq!(swapped, vec!(0x41, 0x42, 0x41, 0x42));
```

### Mutating File-like Data

```rust
use std::io::Cursor;
use bswp::pattern::{Pattern, Predicate};
use bswp::io::swap_io;

// in memory reader (implements `Read`)
let mut reader: Cursor<Vec<u8>> = Cursor::new(vec![0x41, 0x42, 0x43, 0x44]);
// in memory writer (implements `Write`)
let mut writer: Cursor<Vec<u8>> = Cursor::new(Vec::new());

let swaps: &[(Pattern, Predicate)] = &[(Pattern::new(0x42).with_mask(0xFF), Predicate::new().with_periodicity(2).with_offset(0))];
let swap = swap_io(&mut reader, &mut writer, swaps);
assert!(swap.is_ok());
assert_eq!(swap.unwrap(), 4); // 4 bytes written
assert_eq!(writer.into_inner(), vec![0x42, 0x42, 0x42, 0x44])
```

## Changelog

Please see the [CHANGELOG](CHANGELOG.md) for a release history.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.0](https://crates.io/crates/bswp/0.1.0) Mar 27, 2020

### Added

* `Pattern` and `Predicate` builders
* non exhaustive on `Pattern` and `Predicate` structs
* `Pattern` and `Predicate` structs fields now public
* `BytePattern` and `PositionPredicate` traits
* `iter_swap` renamed to `swap_iter`
* generic `swap_iter` and `swap_io`

### Breaking Changes

**Everything** was renamed and modified `^^'`

## [0.1.0](https://crates.io/crates/bswp/0.1.0) Mar 24, 2020

### Added

* `pattern::BytePattern`
* `pattern::Locality`
* `pattern::iter_swap`
* `io::swap_io`


src/lib.rs
//! Utilities for swapping bytes using patterns and masks.
//!
//! # Quick Start
//!
//! ## Iterators
//!
//! ```
//! use bswp::pattern::{Pattern, Predicate, swap_iter};
//!
//! let pattern = Pattern::new(0x42).with_mask(0xFF); // replace byte by 0x42
//! let predicate = Predicate::new().with_periodicity(2).with_offset(1); // replace odd bytes
//! let swaps = &[(pattern, predicate)];
//!
//! let source: [u8; 4] = [0x41, 0x41, 0x41, 0x41];
//! let swapped = swap_iter(&source, swaps); // iterator on result
//! let swapped: Vec<u8> = swapped.collect();
//! assert_eq!(swapped, vec!(0x41, 0x42, 0x41, 0x42));
//! ```
//!
//! ## Mutating File-like Data
//!
//! ```
//! use std::io::Cursor;
//! use bswp::pattern::{Pattern, Predicate};
//! use bswp::io::swap_io;
//!
//! // in memory reader (implements `Read`)
//! let mut reader: Cursor<Vec<u8>> = Cursor::new(vec![0x41, 0x42, 0x43, 0x44]);
//! // in memory writer (implements `Write`)
//! let mut writer: Cursor<Vec<u8>> = Cursor::new(Vec::new());
//!
//! let swaps: &[(Pattern, Predicate)] = &[(Pattern::new(0x42).with_mask(0xFF), Predicate::new().with_periodicity(2).with_offset(0))];
//! let swap = swap_io(&mut reader, &mut writer, swaps);
//! assert!(swap.is_ok());
//! assert_eq!(swap.unwrap(), 4); // 4 bytes written
//! assert_eq!(writer.into_inner(), vec![0x42, 0x42, 0x42, 0x44])
//! ```

/// default buffer size for io: 8KB
pub const BUFFER_SIZE: usize = 8000; // 8KB

/// Predicate on byte position.
pub trait PositionPredicate {
    /// Returns `true` if `position` matches predicate else `false`.
    fn eval(&self, position: usize) -> bool;
}

/// Pattern on byte.
pub trait BytePattern {
    /// Returns the value with current pattern applied.
    fn eval(&self, value: u8) -> u8;
}

pub mod io;
pub mod pattern;


src/pattern.rs
//! Pattern, Predicate and iterator utils (pure)

use crate::{BytePattern, PositionPredicate};

/// Stores a predicate on byte position.
#[non_exhaustive]
pub struct Predicate {
    /// only match every `periodicity` bytes once the `offset` is reached.
    pub periodicity: usize,
    /// skip `offset` first bytes
    /// only bytes with position >= offset will match
    pub offset: usize,
    /// if present only match at most `limit` times bytes.
    ///
    /// it means byte position must verify `(position - offset) / periodicity < limit`
    ///
    /// `None` means no limit.
    pub limit: Option<usize>,
}

/// BytePattern
/// Given a target byte, the pattern will set bits to corresponding bits from `value`
/// according to `mask`.
#[non_exhaustive]
pub struct Pattern {
    pub value: u8,
    pub mask: u8,
}

impl Pattern {
    /// Creates a new `BytePattern` with provided `value`.
    ///
    /// `mask`: 0xFF
    ///
    /// ```
    /// # use bswp::pattern::Pattern;
    /// let byte_pattern = Pattern::new(0xFF).with_mask(0xF0); // replace first 4 bits with 0xF
    /// ```
    pub fn new(value: u8) -> Self {
        Pattern { value, mask: 0xFF }
    }

    pub fn with_mask(mut self, mask: u8) -> Self {
        self.mask = mask;
        self
    }
}

impl BytePattern for Pattern {
    /// Returns the value with current pattern applied.
    ///
    /// ```
    /// use bswp::BytePattern;
    /// use bswp::pattern::Pattern;
    /// let byte_pattern = Pattern::new(0xFF).with_mask(0xF0);
    /// assert_eq!(byte_pattern.eval(0x00), 0xF0);
    /// let byte_pattern = Pattern::new(0b10101111).with_mask(0b10011010);
    /// assert_eq!(byte_pattern.eval(0b00000000), 0b10001010);
    /// ```
    fn eval(&self, value: u8) -> u8 {
        (self.mask & self.value) | (!self.mask & value)
    }
}

impl Default for Predicate {
    /// Creates a default `Predicate`
    ///
    /// `periodicity`: `1` (every bytes)
    /// `offset`: `0` (starting from byte at position `0`)
    /// `limit`: `None` (
    fn default() -> Self {
        Predicate {
            periodicity: 1,
            offset: 0,
            limit: None,
        }
    }
}

impl Predicate {
    /// Creates a new `Predicate` with default `periodicity`, `offset` and `limit`.
    ///
    /// returns `Predicate::default()`
    ///
    /// ```
    /// # use bswp::pattern::Predicate;
    /// # use bswp::PositionPredicate;
    /// let every_bytes = Predicate::new(); // matches every bytes
    /// let even_bytes = Predicate::new().with_periodicity(2); // matches even bytes positions
    /// let odd_bytes = Predicate::new().with_periodicity(2).with_offset(1); // matches odd byte positions
    /// let four_first_bytes = Predicate::new().with_limit(4); // 4 fist bytes
    /// ```
    pub fn new() -> Predicate {
        Predicate::default()
    }

    /// Sets the `periodicity`.
    ///
    /// **Default**: `1`
    pub fn with_periodicity(mut self, periodicity: usize) -> Self {
        self.periodicity = periodicity;
        self
    }

    /// Sets the `offset`.
    ///
    /// **Default**: `0`
    pub fn with_offset(mut self, offset: usize) -> Self {
        self.offset = offset;
        self
    }

    /// Sets the `limit`.
    ///
    /// **Default**: `None`
    pub fn with_limit(mut self, limit: usize) -> Self {
        self.limit = Some(limit);
        self
    }

    /// Sets the `limit` to `None` (Default).
    pub fn with_no_limit(mut self) -> Self {
        self.limit = None;
        self
    }
}

impl PositionPredicate for Predicate {
    /// Returns `true` if `position` matches predicate else `false`.
    ///
    /// ```
    /// # use bswp::pattern::Predicate;
    /// # use bswp::PositionPredicate;
    /// let every_bytes = Predicate::new(); // matches every bytes
    /// assert!(every_bytes.eval(42));
    /// let even_bytes = Predicate::new().with_periodicity(2); // matches even bytes positions
    /// assert!(even_bytes.eval(0) && even_bytes.eval(2));
    /// let odd_bytes = Predicate::new().with_periodicity(2).with_offset(1); // matches odd byte positions
    /// assert!(odd_bytes.eval(1) && odd_bytes.eval(3));
    /// let four_first_bytes = Predicate::new().with_limit(4); // 4 fist bytes
    /// assert!(four_first_bytes.eval(0) && four_first_bytes.eval(2) &&
    ///         four_first_bytes.eval(3) && four_first_bytes.eval(3));
    /// ```
    fn eval(&self, position: usize) -> bool {
        (position >= self.offset)
            && self.limit.map_or(true, |limit| {
                ((position - self.offset) / self.periodicity) < limit
            })
            && ((position - self.offset) % self.periodicity) == 0
    }
}

/// Returns an iterator on swapped bytes from `source`.
///
/// ```
/// use bswp::pattern::{Predicate, Pattern, swap_iter};
/// let predicate = Predicate::new().with_periodicity(2).with_offset(1);
/// let pattern = Pattern::new(0x42).with_mask(0xFF);
/// let swaps = &[(pattern, predicate)];
///
/// let source: [u8; 4] = [0x41, 0x41, 0x41, 0x41];
/// let swapped = swap_iter(&source, swaps);
/// let swapped: Vec<u8> = swapped.collect();
/// assert_eq!(swapped, vec!(0x41, 0x42, 0x41, 0x42));
/// ```
pub fn swap_iter<'a, P: BytePattern, Q: PositionPredicate>(
    source: &'a [u8],
    swaps: &'a [(P, Q)],
) -> impl Iterator<Item = u8> + 'a {
    source.iter().enumerate().map(move |(position, &e)| {
        swaps.iter().fold(e, |value, (pattern, predicate)| {
            if predicate.eval(position) {
                pattern.eval(value)
            } else {
                value
            }
        })
    })
}

#[cfg(test)]
mod tests {
    use crate::pattern::{swap_iter, Pattern, Predicate};
    use crate::{BytePattern, PositionPredicate};

    #[test]
    fn test_swap() {
        let pattern = Pattern::new(0xFF).with_mask(0x0F);
        assert_eq!(pattern.eval(0x0), 0x0F);
        let pattern = Pattern::new(0xFF).with_mask(0xF0);
        assert_eq!(pattern.eval(0x0), 0xF0);
    }

    #[test]
    fn test_eval_2_3_none() {
        let predicate = Predicate::new().with_periodicity(2).with_offset(3);
        let unexpected_vec: Vec<usize> = vec![0, 1, 2, 4, 6, 8];
        let expected_vec: Vec<usize> = vec![3, 5, 7, 9];
        for unexpected in unexpected_vec {
            assert!(!predicate.eval(unexpected));
        }
        for expected in expected_vec {
            assert!(predicate.eval(expected));
        }
    }

    #[test]
    fn test_eval_2_3_2() {
        let predicate = Predicate::new()
            .with_periodicity(2)
            .with_offset(3)
            .with_limit(2);
        let unexpected_vec: Vec<usize> = vec![0, 1, 2, 4, 6, 7, 8, 9]; // 7, 9 unexpected because of limit
        let expected_vec: Vec<usize> = vec![3, 5];
        for unexpected in unexpected_vec {
            assert!(!predicate.eval(unexpected), "{} is unexpected", unexpected);
        }
        for expected in expected_vec {
            assert!(predicate.eval(expected));
        }
    }

    #[test]
    fn test_eval_2_1_none() {
        let predicate = Predicate::new().with_periodicity(2).with_offset(3);
        let unexpected_vec: Vec<usize> = vec![0, 2, 4, 6, 8];
        let expected_vec: Vec<usize> = vec![3, 5, 7, 9];
        for unexpected in unexpected_vec {
            assert!(!predicate.eval(unexpected));
        }
        for expected in expected_vec {
            assert!(predicate.eval(expected));
        }
    }

    #[test]
    fn test_iter_swap() {
        let predicate = Predicate::new().with_periodicity(2).with_offset(1);
        let pattern = Pattern::new(0x42).with_mask(0xFF);
        let swaps = &[(pattern, predicate)];

        let source: [u8; 4] = [0x41, 0x41, 0x41, 0x41];
        let swapped = swap_iter(&source, swaps);
        let swapped: Vec<u8> = swapped.collect();
        assert_eq!(swapped, vec!(0x41, 0x42, 0x41, 0x42));
    }
}


src/io.rs
//! Byte swap IO utils (mut)

use crate::{BytePattern, PositionPredicate, BUFFER_SIZE};
use std::io::{Read, Write};

/// For each byte in `reader` compute pattern and write result to `writer`.
///
/// Returns number of bytes read from `reader` and written to `writer`.
///
/// Internally it relies on a `BUFFER_SIZE` buffer between `reader` and `writer`.
///
/// Please note that `swap_io`:
/// * resets neither of `reader`/`writer` cursors before reading from/writing to it.
/// * resets neither of `reader`/`writer` cursors after reading from/writing to it.
///
/// ```
/// use std::io::Cursor;
/// use bswp::pattern::{Pattern, Predicate};
/// use bswp::io::swap_io;
///
/// // in memory reader (implements `Read`)
/// let mut reader: Cursor<Vec<u8>> = Cursor::new(vec![0x41, 0x42, 0x43, 0x44]);
/// // in memory writer (implements `Write`)
/// let mut writer: Cursor<Vec<u8>> = Cursor::new(Vec::new());
///
/// let swaps: &[(Pattern, Predicate)] = &[(Pattern::new(0x42).with_mask(0xFF), Predicate::new().with_periodicity(2).with_offset(0))];
/// let swap = swap_io(&mut reader, &mut writer, swaps);
/// assert!(swap.is_ok());
/// assert_eq!(swap.unwrap(), 4); // 4 bytes written
/// assert_eq!(writer.into_inner(), vec![0x42, 0x42, 0x42, 0x44])
/// ```
pub fn swap_io<P: BytePattern, Q: PositionPredicate>(
    reader: &mut dyn Read,
    writer: &mut dyn Write,
    swaps: &[(P, Q)],
) -> Result<usize, std::io::Error> {
    let mut position: usize = 0;
    let mut buffer = [0; BUFFER_SIZE];

    loop {
        let size = reader.read(&mut buffer)?;
        if size == 0 {
            break; // finished
        }
        for (position_in_buffer, item) in buffer.iter_mut().enumerate().take(size) {
            let byte_position = position + position_in_buffer; // position relative to reader start
            for (pattern, predicate) in swaps {
                if predicate.eval(byte_position) {
                    *item = pattern.eval(*item);
                }
            }
        }
        position += size;
        writer.write_all(&buffer[..size])?;
    }
    Ok(position)
}
